File: pset6/credit.py
from cs50 import *
import re

while True:
    number = get_int("Number:")
    if number > 0:
        break
digit = len(str(number))
digitArray = list(str(number))  # 注意這邊儲存的是string的 list 所以要做計算時一定要做轉換
odd = 0
even = 0
#  檢查位數是否正確
if digit in [13, 15, 16, 3]:
    # 檢查開頭
    if(re.search("^5[1-5]", str(number)) or re.search("^34|37", str(number)) or re.search("4", str(number))):
        # 跑Luhn算法，這邊注意，i是在digitArray裡面loop，
        # 所以i會等於 digitArray[0,1,2..]的數，因此如果for內用digitArray[i]處理會報錯
        for i in digitArray[digit - 2: -1: -2]:
            print("in1")
            odd = odd + (int(i) * 2 // 10) + (int(i) * 2 % 10)
        for j in digitArray[digit - 1: -1: -2]:
            print("in2")
            even = even + int(j)
        # 跑Luhn算法
        total = odd + even
        if total % 10 == 0:
            # 最後檢查屬於哪個卡 print出結果
            if digit == 15:
                print("AMEX")
            elif digit == 13 or (digit == 16 and int(digitArray[0]) == 4):  # 因為digitArray是string 所以要比對要轉成int
                print("VISA")
            elif digit == 16 and digitArray[0] == '5':  # 或者不轉用'5'也可以
                print("MASTERCARD")
            else:
                print("INVALID")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")

File: pset1/pset1/credit/credit.c
# include <stdio.h>
# include <cs50.h>
# include <math.h>

int main(void)
{
// for user enter number
    long number;

    do
    {
        number = get_long("Number:");
    }
    while (number < 0);

// for checking algorithm digits
    //calculate digit
    int digit = 1;
    long Calculatedigit = number;

    while (Calculatedigit / 10 > 0)
    {
        Calculatedigit = Calculatedigit / 10;
        digit = digit + 1;
    }
    //printf("%d\n", digit);
    
    //check if digit fit credit card digit
    if (digit == 13 || digit == 15 || digit == 16)
    {
        if (number / 10000000000000 == 34 || number / 10000000000000 == 37 || number / 1000000000000 == 4 || number / 1000000000000000 == 4 
            || number / 100000000000000 == 51 || number / 100000000000000 == 52 || number / 100000000000000 == 53 
            || number / 100000000000000 == 54 || number / 100000000000000 == 55)
        {
            int digitArray[digit];

            //save each digit into array
            for (int i = 0; i < digit; i++)
            {
                digitArray[i] = number % 10;
                number = number / 10;
                //printf("%d,", digitArray[i]);
            }

            int even = 0;

            for (int j = 1; j < digit; j = j + 2)
            {
                even = even + (2 * digitArray[j] / 10) + (2 * digitArray[j] % 10);
            }
            //printf("%d\n", even);
            
            int odd = 0;
            for (int k = 0; k < digit; k = k + 2)
            {
                odd = odd + digitArray[k];
                //printf("%d,",digitArray[k]);
            }
            //printf("%d\n", odd);
            int checksum = 0;
            checksum = odd + even;
            if (checksum % 10  == 0)
            {
                if (digit == 15)
                {
                    printf("AMEX\n");
                }
                else if (digit == 13)
                {
                    printf("VISA\n");
                }
                else if (digit == 16 && digitArray[15] == 4)
                {
                    printf("VISA\n");
                }
                else if (digit == 16 && digitArray[15] == 5)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }


    }
    else
    {
        printf("INVALID\n");
    }
}
